crates/extensions/c8y_mapper_ext/src/config_operations.rs

```diff
@@ -0,0 +1,1253 @@
+    [
+        mqtt_schema.topics(AnyEntity, Command(OperationType::ConfigUpdate)),
+        mqtt_schema.topics(AnyEntity, CommandMetadata(OperationType::ConfigUpdate)),
+    ]
+    .into_iter()
+    .collect()
+}
+
+impl CumulocityConverter {
+    /// Convert a SmartREST request to a ThinEdge config_snapshot command
+    pub fn convert_config_snapshot_request(
+        &self,
+        smartrest: &str,
+    ) -> Result<Vec<Message>, CumulocityMapperError> {
+        if !self.config.capabilities.config_snapshot {
+            // config_snapshot is disabled
+            return Ok(vec![]);
+        }
+
+        let snapshot_request = SmartRestConfigUploadRequest::from_smartrest(smartrest)?;
+        let target = self
+            .entity_store
+            .get_by_external_id(&snapshot_request.device.clone().into())
+            .ok_or_else(|| UnknownDevice {
+                device_id: snapshot_request.device.to_string(),
+            })?;
+
+        let cmd_id = nanoid!();
+        let channel = Channel::Command {
+            operation: OperationType::ConfigSnapshot,
+            cmd_id: cmd_id.clone(),
+        };
+        let topic = self.mqtt_schema.topic_for(&target.topic_id, &channel);
+        let external_id: String = target.external_id.clone().into();
+
+        let tedge_url = format!(
+            "http://{}/tedge/file-transfer/{}/config_snapshot/{}-{}",
+            &self.config.tedge_http_host, external_id, snapshot_request.config_type, cmd_id
+        );
+
+        let request = ConfigSnapshotCmdPayload {
+            status: CommandStatus::Init,
+            tedge_url,
+            config_type: snapshot_request.config_type,
+            path: None,
+        };
+
+        // Command messages must be retained
+        Ok(vec![Message::new(&topic, request.to_json()).with_retain()])
+    }
+
+    /// Address a received ThinEdge config_snapshot command. If its status is
+    /// - "executing", it converts the message to SmartREST "Executing".
+    /// - "successful", it uploads a config snapshot to c8y and converts the message to SmartREST "Successful".
+    /// - "failed", it converts the message to SmartREST "Failed".
+    pub async fn handle_config_snapshot_state_change(
+        &mut self,
+        topic_id: &EntityTopicId,
+        cmd_id: &str,
+        message: &Message,
+    ) -> Result<Vec<Message>, ConversionError> {
+        if !self.config.capabilities.config_snapshot {
+            // config_snapshot is disabled
+            return Ok(vec![]);
+        }
+
+        // get the device metadata from its id
+        let device = self.entity_store.get(topic_id).ok_or_else(|| {
+            CumulocityMapperError::UnregisteredDevice {
+                topic_id: topic_id.to_string(),
+            }
+        })?;
+        let external_id = &device.external_id;
+
+        let smartrest_topic = C8yTopic::smartrest_response_topic(device).unwrap(); // later fix
```

Comment:
```suggestion
        let smartrest_topic = self.smartrest_publish_topic_for_entity(topic_id);
```
It covers hierarchical devices as well.